htp: fix case of nested FormField if first formField usage is with single parameter

jrudolph opened the pull request:
So far, in #2283 the case for tuples was fixed which also worked for nested `formField` directives in most of the cases but not when the first usage was with a single argument.

Refs #73.

johanandren commented:
Tried to review this and looked at the previous PR but I don't understand how it works, could you give a brief explanation @jrudolph ?

jrudolph commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/FormFieldDirectives.scala:
```diff
@@ -52,22 +55,35 @@ trait FormFieldDirectives extends ToNameReceptacleEnhancements {
    */
-  def formFields(pdm: FieldMagnet): pdm.Out = pdm()
+  override def formFields(pdm: FieldMagnet): Directive[pdm.U] =
+    toStrictEntity(StrictForm.toStrictTimeout).wrap { pdm() }
```
That's basically the fix. We want to be sure that the complete inner part of a formField directive is wrapped in a `toStrictEntity` directive.

The problem in the old version was, that we couldn't prove that `pdm()` actually is a directive, so we couldn't use `wrap`.

jrudolph commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/FormFieldDirectives.scala:
```diff
@@ -206,25 +223,29 @@ object FormFieldDirectives extends FormFieldDirectives {
     import akka.http.scaladsl.server.util.TupleOps._
 
-    implicit def forTuple[T](implicit fold: FoldLeft[Directive0, T, ConvertFieldDefAndConcatenate.type]): FieldDefAux[T, fold.Out] =
-      fieldDef[T, fold.Out](fold(toStrictEntity(StrictForm.toStrictTimeout), _))
```
That's the old fix which fixed it for `formField('a, 'b)`, i.e. if there are more than two form fields, than fold the directives starting with `toStrictEntity` on the outside.

It didn't fix it for the other cases of a single form field, though. One potential solution would be to add `toStrictEntity` to all the implicits for the single form field above, that would work but would mean that we would add `toStrictEntity` as many times as we have `formFields` which would add a bit of overhead.

jrudolph replied:
Thinking about it, there's already existing overhead: for every given field, we run the complete unmarshalling which seems wasteful. So, that could be another area of improvement.

jrudolph commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/FormFieldDirectives.scala:
```diff
@@ -113,27 +129,28 @@ object FormFieldDirectives extends FormFieldDirectives {
-    type Out
-    def apply(): Out
+    type U
+    def apply(): Directive[U]
```
So, the typesafe fix is to add enough types here (and everywhere) to declare that the result of `pdf()` above actually is a `Directive` and not just anything.

That is we add the `Directive` type here and remove the `Directive` wrapping in all types below. Note that `type Directive1[T] = Directive[Tuple1[T]]` etc.

raboof commented:
MiMa:
```
[error]  * method formFields(akka.http.scaladsl.server.directives.FormFieldDirectives#FieldMagnet)java.lang.Object in trait akka.http.scaladsl.server.directives.FormFieldDirectives does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.server.directives.FormFieldDirectives.formFields")
[error]  * method formField(akka.http.scaladsl.server.directives.FormFieldDirectives#FieldMagnet)java.lang.Object in trait akka.http.scaladsl.server.directives.FormFieldDirectives does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.server.directives.FormFieldDirectives.formField")
[error]  * method formFields(akka.http.scaladsl.server.directives.FormFieldDirectives#FieldMagnet)akka.http.scaladsl.server.Directive in trait akka.http.scaladsl.server.directives.FormFieldDirectives is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.server.directives.FormFieldDirectives.formFields")
[error]  * method formField(akka.http.scaladsl.server.directives.FormFieldDirectives#FieldMagnet)akka.http.scaladsl.server.Directive in trait akka.http.scaladsl.server.directives.FormFieldDirectives is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]
```

jrudolph commented:
> MiMa:

These mima failures happen only on 2.11. I tried to keep bin compat by adding this FormFieldDirectivesCompat subclass but that probably won't help for 2.11. I currently don't know how to solve that problem.

jrudolph commented:
> MiMa:

> These mima failures happen only on 2.11. I tried to keep bin compat by adding this FormFieldDirectivesCompat subclass but that probably won't help for 2.11. I currently don't know how to solve that problem.

Most easily by changing the return type to a dummy type member similar to before so that the erased result type is still `Object`/`AnyRef`.

raboof commented:
Hmm tricky.

Perhaps we should start planning dropping support for 2.11 (some time after Akka 2.6 is released?) and postponing this PR until then?

jrudolph commented:
I think I fixed the incompatibility now. The idea was to keep the `Out` type member alongside the new `U` type member. This way the erased result type is still `AnyRef`. I added different alternative using `@pre213` and `@since213` so that we can drop the extra structure once 2.12 support is gone.
